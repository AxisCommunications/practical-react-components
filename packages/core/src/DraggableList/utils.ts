import React from 'react'

export const range = (length: number) => [...Array(length).keys()]

export const move = (
  arrRange: ReadonlyArray<number>,
  locked: ReadonlyArray<boolean>,
  from: number,
  to: number
) => {
  let p = 0
  return arrRange.map(i => {
    if (locked[i]) {
      return i
    }
    if (i === to) {
      return from
    }
    while (locked[p] || p === from) p++
    return p++
  })
}

export const isElementDisabled = <T>(el: T) =>
  React.isValidElement<{ disabled: boolean }>(el)
    ? el.props.disabled === true
    : false
