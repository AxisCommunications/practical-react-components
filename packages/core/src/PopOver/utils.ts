import { useLayoutEffect } from 'react'

// Reposition on scroll
export const useOnScrollEffect = (
  anchorEl: HTMLElement | null,
  onScroll: () => void
): void => {
  useLayoutEffect(() => {
    if (anchorEl === null) {
      return undefined
    }

    let parentEl = anchorEl.parentElement
    const scrollableParents: Array<HTMLElement> = []
    while (parentEl !== null) {
      if (parentEl.scrollHeight > parentEl.clientHeight) {
        scrollableParents.push(parentEl)
      }

      parentEl = parentEl.parentElement
    }
    scrollableParents.forEach(parent => {
      parent.addEventListener('scroll', onScroll)
    })
    window.addEventListener('scroll', onScroll)
    return () => {
      scrollableParents.forEach(parent => {
        parent.removeEventListener('scroll', onScroll)
      })
      window.removeEventListener('scroll', onScroll)
    }
  }, [onScroll, anchorEl])
}

// Reposition on resize
export const useOnResizeEffect = (
  anchorEl: HTMLElement | null,
  onResize: () => void
): void => {
  useLayoutEffect(() => {
    if (anchorEl === null) {
      return undefined
    }

    let parentEl = anchorEl.parentElement
    const observer = new window.ResizeObserver(onResize)

    while (parentEl !== null) {
      observer.observe(parentEl)
      parentEl = parentEl.parentElement
    }

    return () => observer.disconnect()
  }, [onResize, anchorEl])
}

export const anchorPosition = (
  anchorEl: HTMLElement,
  popOverContainerEl: HTMLElement,
  {
    horizontalPosition = 'left',
    horizontalAlignment = 'left',
    verticalPosition = 'bottom',
    verticalAlignment = 'top',
  }: {
    readonly horizontalPosition?: 'left' | 'right' | 'center'
    readonly horizontalAlignment?: 'left' | 'right' | 'center'
    readonly verticalPosition?: 'top' | 'bottom' | 'center'
    readonly verticalAlignment?: 'top' | 'bottom' | 'center'
  }
) => {
  // ðŸŽ“ Set up default positioning
  const anchorBBox = anchorEl.getBoundingClientRect()
  const { top, left, bottom, right, width, height } = anchorBBox
  // Ignore invisible anchor, this might happen when the anchor's
  // visibility changes with e.g. display = 'none', in which case the
  // popover will get the wrong position.
  if (width === 0 && height === 0) {
    return
  }
  const anchorHPos = { left, right, center: (left + right) / 2 }
  const anchorVPos = { top, bottom, center: (top + bottom) / 2 }

  const popOverBBox = popOverContainerEl.getBoundingClientRect()
  const popOverHOffset = {
    left: 0,
    right: -popOverBBox.width,
    center: -popOverBBox.width / 2,
  }
  const popOverVOffset = {
    top: 0,
    bottom: -popOverBBox.height,
    center: -popOverBBox.height / 2,
  }

  let popOverLeft =
    anchorHPos[horizontalPosition] + popOverHOffset[horizontalAlignment]
  let popOverTop =
    anchorVPos[verticalPosition] + popOverVOffset[verticalAlignment]

  // ðŸ§š Automagically keep the pop-over within the document boundaries,
  // by repositioning it if it would overflow.
  const { clientHeight, clientWidth } = document.documentElement
  const popOverVOverflow = clientHeight - (popOverTop + popOverBBox.height)
  const popOverHOverflow = clientWidth - (popOverLeft + popOverBBox.width)
  if (popOverVOverflow < 0) {
    popOverTop = popOverTop + popOverVOverflow
  }
  if (popOverHOverflow < 0) {
    popOverLeft = popOverLeft + popOverHOverflow
  }

  // Set final positioning on pop-over's style
  popOverContainerEl.style.top = `${popOverTop}px`
  popOverContainerEl.style.left = `${popOverLeft}px`
}
