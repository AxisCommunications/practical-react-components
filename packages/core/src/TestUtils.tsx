import React from 'react'
import renderer from 'react-test-renderer'
import 'jest-styled-components'
import { PracticalProvider } from './Practical'

/*
  Only used for Jest Snapshot Testing
*/

export const TP = ({
  children,
}: {
  readonly children: React.ReactElement
}): JSX.Element => <PracticalProvider>{children}</PracticalProvider>

/**
 * Mock reference to practical-root
 */
const createNodeMock = (element: React.ReactElement) => {
  if (element.props.id !== 'practical-root') {
    return null
  }

  return {}
}

export const TestRender = (jsx: JSX.Element) => {
  return renderer.create(<TP>{jsx}</TP>, { createNodeMock }).toJSON()
}

export const TestText = (): JSX.Element => <span>Test</span>

/* eslint-disable-next-line react/display-name */
export const TestTextWithForwardRef = React.forwardRef<HTMLElement>(
  ({ ...props }, ref) => (
    <span {...props} ref={ref}>
      Test
    </span>
  )
)
