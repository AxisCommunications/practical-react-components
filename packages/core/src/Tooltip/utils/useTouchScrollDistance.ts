import { useState, useEffect, useLayoutEffect } from 'react'

export const useTouchScrollDistance = () => {
  const [origin, setOrigin] = useState<TouchList | null>(null)
  const [touches, setTouches] = useState<TouchList | null>(null)
  /**
   * The distance between touch origin and touch current for both
   * x-axis and y-axis
   */
  const [touchScrollDistance, setTouchScrollDistance] = useState({ x: 0, y: 0 })

  useEffect(() => {
    const touchStartHandler = (event: TouchEvent) => {
      if (origin === null) {
        setOrigin(event.touches)
      }
    }

    const touchMoveHandler = (event: TouchEvent) =>
      setTouches(event.changedTouches)

    const touchEndHandler = (event: TouchEvent) => {
      if (event.touches.length === 0) {
        setOrigin(null)
        setTouches(null)
        setTouchScrollDistance({ x: 0, y: 0 })
      }
    }

    const touchCancelHandler = () => {
      setOrigin(null)
      setTouches(null)
      setTouchScrollDistance({ x: 0, y: 0 })
    }

    document.addEventListener('touchstart', touchStartHandler)
    document.addEventListener('touchmove', touchMoveHandler)
    document.addEventListener('touchend', touchEndHandler)
    document.addEventListener('touchcancel', touchCancelHandler)

    return () => {
      document.removeEventListener('touchstart', touchStartHandler)
      document.removeEventListener('touchmove', touchMoveHandler)
      document.removeEventListener('touchend', touchEndHandler)
      document.removeEventListener('touchcancel', touchCancelHandler)
    }
  }, [origin])

  /**
   * Calculates the distance in pixels between the origin of
   * a touch event and position updates to that touch event.
   */
  useLayoutEffect(() => {
    if (origin === null || touches === null) {
      return
    }

    // User is not scrolling
    if (touches.length > 1) {
      return
    }

    const deltaX = touches[0].clientX - origin[0].clientX
    const deltaY = touches[0].clientY - origin[0].clientY

    setTouchScrollDistance({ x: deltaX, y: deltaY })
  }, [origin, touches])

  return touchScrollDistance
}
