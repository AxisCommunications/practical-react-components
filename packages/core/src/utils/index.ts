export * from './withField'
export * from './twoLinesClamp'
