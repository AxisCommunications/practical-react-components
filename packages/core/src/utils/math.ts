/**
 * Positive remainder operator
 *
 * @param dividend
 * @param divisor
 */
export const remainder = (dividend: number, divisor: number) =>
  ((dividend % divisor) + divisor) % divisor
