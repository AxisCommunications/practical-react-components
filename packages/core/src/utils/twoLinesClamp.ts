import { css } from 'styled-components'

/**
 * A styled-components mixin that truncates two-lines text with ellipsis.
 *
 * An example:
 * ```tsx
 * const Div = styled.div`
 *   ${twoLinesClamp};
 * `
 * ```
 */
export const twoLinesClamp = css`
  /* stylelint-disable-next-line value-no-vendor-prefix */
  display: -webkit-box;
  /* stylelint-disable-next-line property-no-vendor-prefix */
  -webkit-line-clamp: 2;
  /* stylelint-disable-next-line property-no-vendor-prefix */
  -webkit-box-orient: vertical;
  /* stylelint-disable-next-line property-no-vendor-prefix */
  -moz-box-orient: vertical;
  overflow: hidden;
  white-space: pre-wrap;
`
