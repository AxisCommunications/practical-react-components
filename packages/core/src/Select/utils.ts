import { remainder } from '../utils/math'

/**
 * TODO: Fix width according to UX
 */
export const getWidth = (width: string): string => {
  switch (width) {
    case 'small':
      return '160px'
    case 'medium':
      return '220px'
    case 'large':
      return '300px'
    case 'full':
      return '100%'
    default:
      return '100%'
  }
}

/**
 * A scrollIntoView with fallback for Edge. For browsers that support it
 * this is identical to:
 *
 * `(element) => element.scrollIntoView({ block: 'nearest' })
 *
 * For other browser this is implemented with `element.scrollIntoView(true)`
 * or `element.scrollIntoView(false)` depending on the vertical offset. It
 * doesn't polyfill the full functionality of `scrollIntoView`.
 */
export const scrollIntoView = (element: Element) => {
  if ('scrollBehavior' in window.document.documentElement.style) {
    element.scrollIntoView({ block: 'nearest' })
    return
  }

  const parent = element.parentElement as Element
  const { height, top }: ClientRect = element.getBoundingClientRect()
  const { top: parentTop }: ClientRect = parent.getBoundingClientRect()
  const offset = top - parentTop

  if (offset < 0) {
    element.scrollIntoView(true)
  } else if (offset + height > parent.clientHeight) {
    element.scrollIntoView(false)
  }
}

type Options = ReadonlyArray<{ readonly disabled?: boolean }>

export const findNextIndex = (index: number | null, options: Options) => {
  const start = index === null ? 0 : index + 1
  for (let i = start; i < start + options.length; i++) {
    const nextIndex = remainder(i, options.length)
    if (options[nextIndex].disabled !== true) {
      return nextIndex
    }
  }
  return index
}

export const findPrevIndex = (index: number | null, options: Options) => {
  const start = index === null ? options.length - 1 : index - 1
  for (let i = start; i > start - options.length; i--) {
    const nextIndex = remainder(i, options.length)
    if (options[nextIndex].disabled !== true) {
      return nextIndex
    }
  }
  return index
}
