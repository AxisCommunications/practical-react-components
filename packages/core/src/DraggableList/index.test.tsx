import React from 'react'
import 'jest-styled-components'

import { DraggableList, DraggableListItem } from '.'
import { move } from './utils'
import { TestRender } from '../TestUtils'

const callback = () => {
  /** */
}

describe('DraggableList', () => {
  test('DraggableList', () => {
    const tree = TestRender(
      <DraggableList onChange={callback}>
        {[1, 2, 3, 4, 5].map((item, index) => (
          <DraggableListItem key={index} disabled={item === 4}>
            <p>{item}</p>
          </DraggableListItem>
        ))}
      </DraggableList>
    )
    expect(tree).toMatchSnapshot()
  })

  test('move', () => {
    const testItems = [
      {
        arr: [0, 1, 2, 3, 4],
        locked: [false, false, true, false, false],
        from: 3,
        to: 1,
        resultIndices: [0, 3, 2, 1, 4],
      },
      {
        arr: [0, 1, 2, 3, 4],
        locked: [false, false, true, true, false],
        from: 4,
        to: 1,
        resultIndices: [0, 4, 2, 3, 1],
      },
      {
        arr: [0, 1, 2, 3, 4],
        locked: [false, false, true, true, false],
        from: 1,
        to: 4,
        resultIndices: [0, 4, 2, 3, 1],
      },
      {
        arr: [0, 1, 2, 3, 4],
        locked: [true, false, false, false, true],
        from: 1,
        to: 3,
        resultIndices: [0, 2, 3, 1, 4],
      },
    ]

    testItems.forEach(({ arr, locked, from, to, resultIndices }) =>
      expect(move(arr, locked, from, to)).toEqual(resultIndices)
    )
  })
})
