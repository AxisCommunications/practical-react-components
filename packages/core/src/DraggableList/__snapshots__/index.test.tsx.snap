// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DraggableList DraggableList 1`] = `
.c6 {
  height: 1px;
  background-color: rgb(230,230,230);
}

.c5 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  pointer-events: none;
  overflow: hidden;
  vertical-align: middle;
  height: 24px;
  width: 24px;
}

.c5 > svg {
  fill: currentColor;
  height: 100%;
  width: auto;
}

.c2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
  width: 100%;
  height: 48px;
}

.c4 {
  color: rgb(102,102,102);
  cursor: move;
}

.c3 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  height: 100%;
  -webkit-box-flex: 1;
  -webkit-flex-grow: 1;
  -ms-flex-positive: 1;
  flex-grow: 1;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.c8 {
  position: fixed;
  top: 0;
  width: 360px;
  border-radius: 4px;
  margin: 8px 16px;
  right: 0;
}

.c0 {
  color: rgb(41,41,41);
}

.c1 {
  position: absolute;
  z-index: 0;
}

.c7 {
  position: absolute;
  z-index: 1;
}

<div
  className="c0"
  id="practical-root"
>
  <div>
    <div
      className="c1"
      id="layer-1"
    >
      <div>
        <div>
          <div
            className="c2"
            data-index={0}
            onDragEnd={[Function]}
            onDragOver={[Function]}
            onDragStart={[Function]}
            style={
              Object {
                "opacity": undefined,
              }
            }
          >
            <div
              className="c3"
              disabled={false}
            >
              <p>
                1
              </p>
            </div>
            <div
              className="c4"
              onMouseDown={[Function]}
              onMouseUp={[Function]}
            >
              <span
                className="c5"
                role="img"
                size="medium"
              >
                <svg
                  height="1em"
                  viewBox="0 0 24 24"
                  width="1em"
                >
                  <path
                    d="m20 9h-16v2h16v-2zm-16 6h16v-2h-16v2z"
                  />
                </svg>
              </span>
            </div>
          </div>
          <div
            className="c6"
          />
        </div>
        <div>
          <div
            className="c2"
            data-index={1}
            onDragEnd={[Function]}
            onDragOver={[Function]}
            onDragStart={[Function]}
            style={
              Object {
                "opacity": undefined,
              }
            }
          >
            <div
              className="c3"
              disabled={false}
            >
              <p>
                2
              </p>
            </div>
            <div
              className="c4"
              onMouseDown={[Function]}
              onMouseUp={[Function]}
            >
              <span
                className="c5"
                role="img"
                size="medium"
              >
                <svg
                  height="1em"
                  viewBox="0 0 24 24"
                  width="1em"
                >
                  <path
                    d="m20 9h-16v2h16v-2zm-16 6h16v-2h-16v2z"
                  />
                </svg>
              </span>
            </div>
          </div>
          <div
            className="c6"
          />
        </div>
        <div>
          <div
            className="c2"
            data-index={2}
            onDragEnd={[Function]}
            onDragOver={[Function]}
            onDragStart={[Function]}
            style={
              Object {
                "opacity": undefined,
              }
            }
          >
            <div
              className="c3"
              disabled={false}
            >
              <p>
                3
              </p>
            </div>
            <div
              className="c4"
              onMouseDown={[Function]}
              onMouseUp={[Function]}
            >
              <span
                className="c5"
                role="img"
                size="medium"
              >
                <svg
                  height="1em"
                  viewBox="0 0 24 24"
                  width="1em"
                >
                  <path
                    d="m20 9h-16v2h16v-2zm-16 6h16v-2h-16v2z"
                  />
                </svg>
              </span>
            </div>
          </div>
          <div
            className="c6"
          />
        </div>
        <div>
          <div
            className="c2"
            data-index={3}
            onDragEnd={[Function]}
            onDragStart={[Function]}
            style={
              Object {
                "opacity": undefined,
              }
            }
          >
            <div
              className="c3"
              disabled={true}
            >
              <p>
                4
              </p>
            </div>
          </div>
          <div
            className="c6"
          />
        </div>
        <div>
          <div
            className="c2"
            data-index={4}
            onDragEnd={[Function]}
            onDragOver={[Function]}
            onDragStart={[Function]}
            style={
              Object {
                "opacity": undefined,
              }
            }
          >
            <div
              className="c3"
              disabled={false}
            >
              <p>
                5
              </p>
            </div>
            <div
              className="c4"
              onMouseDown={[Function]}
              onMouseUp={[Function]}
            >
              <span
                className="c5"
                role="img"
                size="medium"
              >
                <svg
                  height="1em"
                  viewBox="0 0 24 24"
                  width="1em"
                >
                  <path
                    d="m20 9h-16v2h16v-2zm-16 6h16v-2h-16v2z"
                  />
                </svg>
              </span>
            </div>
          </div>
          <div
            className="c6"
          />
        </div>
      </div>
    </div>
  </div>
  <div>
    <div
      className="c7"
      id="layer-2"
    >
      <div
        className="c8"
      />
    </div>
  </div>
</div>
`;
